# Import datasets

In [1]:
#Import pandas, and read the train file
import pandas as pd

train = pd.read_csv('train.csv')

In [2]:
#Check how many cols and rows the data has
print(train.shape)

(1460, 81)


In [3]:
#Check the first five rows of the data
train.head()

Id  MSSubClass MSZoning  LotFrontage  LotArea Street Alley LotShape  \
0   1          60       RL         65.0     8450   Pave   NaN      Reg   
1   2          20       RL         80.0     9600   Pave   NaN      Reg   
2   3          60       RL         68.0    11250   Pave   NaN      IR1   
3   4          70       RL         60.0     9550   Pave   NaN      IR1   
4   5          60       RL         84.0    14260   Pave   NaN      IR1   

  LandContour Utilities  ... PoolArea PoolQC Fence MiscFeature MiscVal MoSold  \
0         Lvl    AllPub  ...        0    NaN   NaN         NaN       0      2   
1         Lvl    AllPub  ...        0    NaN   NaN         NaN       0      5   
2         Lvl    AllPub  ...        0    NaN   NaN         NaN       0      9   
3         Lvl    AllPub  ...        0    NaN   NaN         NaN       0      2   
4         Lvl    AllPub  ...        0    NaN   NaN         NaN       0     12   

  YrSold  SaleType  SaleCondition  SalePrice  
0   2008        WD         Normal     208500  
1   2007        WD         Normal     181500  
2   2008        WD         Normal     223500  
3   2006        WD        Abnorml     140000  
4   2008        WD         Normal     250000  

[5 rows x 81 columns]

In [4]:
#Check a summary of the data
train.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 1460 entries, 0 to 1459
Data columns (total 81 columns):
 #   Column         Non-Null Count  Dtype  
---  ------         --------------  -----  
 0   Id             1460 non-null   int64  
 1   MSSubClass     1460 non-null   int64  
 2   MSZoning       1460 non-null   object 
 3   LotFrontage    1201 non-null   float64
 4   LotArea        1460 non-null   int64  
 5   Street         1460 non-null   object 
 6   Alley          91 non-null     object 
 7   LotShape       1460 non-null   object 
 8   LandContour    1460 non-null   object 
 9   Utilities      1460 non-null   object 
 10  LotConfig      1460 non-null   object 
 11  LandSlope      1460 non-null   object 
 12  Neighborhood   1460 non-null   object 
 13  Condition1     1460 non-null   object 
 14  Condition2     1460 non-null   object 
 15  BldgType       1460 non-null   object 
 16  HouseStyle     1460 non-null   object 
 17  OverallQual    1460 non-null   int64  
 18  OverallC

# Explore The Data

## Number Of Missing Values By Column

In [5]:
#Count (using `.sum()`) the number of missing values (`.isnull()`) in each column of 
#the train data as well as the percentages (using `.mean()` instead of `.sum()`).
#Order them (increasing or decreasing) using sort_values
#Call `pd.concat` to present these in a single table (DataFrame) with the helpful column names 'count' and '%'
missing = pd.concat([train.isnull().sum(), 100 * train.isnull().mean()], axis=1)
missing.columns=['count', '%']
missing.sort_values(by='count')

count          %
Id                 0   0.000000
TotalBsmtSF        0   0.000000
Heating            0   0.000000
SaleCondition      0   0.000000
CentralAir         0   0.000000
...              ...        ...
FireplaceQu      690  47.260274
Fence           1179  80.753425
Alley           1369  93.767123
MiscFeature     1406  96.301370
PoolQC          1453  99.520548

[81 rows x 2 columns]

In [6]:
missing.loc[missing['count']>0].sort_values(by='count')

count          %
Electrical        1   0.068493
MasVnrType        8   0.547945
MasVnrArea        8   0.547945
BsmtQual         37   2.534247
BsmtCond         37   2.534247
BsmtFinType1     37   2.534247
BsmtExposure     38   2.602740
BsmtFinType2     38   2.602740
GarageCond       81   5.547945
GarageQual       81   5.547945
GarageFinish     81   5.547945
GarageType       81   5.547945
GarageYrBlt      81   5.547945
LotFrontage     259  17.739726
FireplaceQu     690  47.260274
Fence          1179  80.753425
Alley          1369  93.767123
MiscFeature    1406  96.301370
PoolQC         1453  99.520548

In [7]:
train['Electrical'].value_counts(), train['Electrical'].value_counts().sum()

(SBrkr    1334
 FuseA      94
 FuseF      27
 FuseP       3
 Mix         1
 Name: Electrical, dtype: int64,
 1459)

###### Description
Electrical: Electrical system

       SBrkr	Standard Circuit Breakers & Romex
       FuseA	Fuse Box over 60 AMP and all Romex wiring (Average)	
       FuseF	60 AMP Fuse Box and mostly Romex wiring (Fair)
       FuseP	60 AMP Fuse Box and mostly knob & tube wiring (poor)
       Mix	Mixed

In [73]:
m = train.loc[train['Electrical'].isnull()]
m['Electrical']

1379    NaN
Name: Electrical, dtype: object

In [74]:
train.iloc[1379, 40:45]

HeatingQC      Gd
CentralAir      Y
Electrical    NaN
1stFlrSF      754
2ndFlrSF      640
Name: 1379, dtype: object

The missing value is marked as 'NA' in the csv file and is what truly missed.

In [8]:
train['MasVnrType'].value_counts(), train['MasVnrType'].value_counts().sum()

(None       864
 BrkFace    445
 Stone      128
 BrkCmn      15
 Name: MasVnrType, dtype: int64,
 1452)

###### Description
MasVnrType: Masonry veneer type

       BrkCmn	Brick Common
       BrkFace	Brick Face
       CBlock	Cinder Block
       None	 None
       Stone	Stone

In [48]:
m = train.loc[train['MasVnrType'].isnull()]
m['MasVnrType']

234     NaN
529     NaN
650     NaN
936     NaN
973     NaN
977     NaN
1243    NaN
1278    NaN
Name: MasVnrType, dtype: object

In [68]:
train.iloc[[234, 529, 650], 23:27]

Exterior1st Exterior2nd MasVnrType  MasVnrArea
234     VinylSd     VinylSd        NaN         NaN
529     Wd Sdng       Stone        NaN         NaN
650     CemntBd     CmentBd        NaN         NaN

The missing values are makred as 'NA' in the csv file and are what truly missed.

In [9]:
train['MasVnrArea'].value_counts(), train['MasVnrArea'].value_counts().sum()

(0.0      861
 72.0       8
 180.0      8
 108.0      8
 120.0      7
         ... 
 651.0      1
 337.0      1
 415.0      1
 293.0      1
 621.0      1
 Name: MasVnrArea, Length: 327, dtype: int64,
 1452)

In [49]:
m = train.loc[train['MasVnrArea'].isnull()]
m['MasVnrArea']

234    NaN
529    NaN
650    NaN
936    NaN
973    NaN
977    NaN
1243   NaN
1278   NaN
Name: MasVnrArea, dtype: float64

There are many unique values in this column and we need to categorize them if want to use for analysis, but this column doesn't look deeply related to the house price so I'll skip it for now.

In addition, there 864 of 'None' in 'MasVnrType' but there are 861 of '0' in 'MasVnrArea'. We need to investigate why there are 3 different data which looks that should be same.

In [10]:
train['BsmtQual'].value_counts(), train['BsmtQual'].value_counts().sum()

(TA    649
 Gd    618
 Ex    121
 Fa     35
 Name: BsmtQual, dtype: int64,
 1423)

BsmtQual: Evaluates the height of the basement

       Ex	Excellent (100+ inches)	
       Gd	Good (90-99 inches)
       TA	Typical (80-89 inches)
       Fa	Fair (70-79 inches)
       Po	Poor (<70 inches
       NA	No Basement

In [52]:
m = train.loc[train['BsmtQual'].isnull()]
m['BsmtQual']

17      NaN
39      NaN
90      NaN
102     NaN
156     NaN
182     NaN
259     NaN
342     NaN
362     NaN
371     NaN
392     NaN
520     NaN
532     NaN
533     NaN
553     NaN
646     NaN
705     NaN
736     NaN
749     NaN
778     NaN
868     NaN
894     NaN
897     NaN
984     NaN
1000    NaN
1011    NaN
1035    NaN
1045    NaN
1048    NaN
1049    NaN
1090    NaN
1179    NaN
1216    NaN
1218    NaN
1232    NaN
1321    NaN
1412    NaN
Name: BsmtQual, dtype: object

In [57]:
train.iloc[17, 30]

nan

NA desn't actually mean missing values, just 'No basement'.
I will replace 'NA' to 'Nb' to prevent misleading the data.

In [27]:
train['BsmtQual'].replace('NA', 'Nb') # didn't work

0       Gd
1       Gd
2       Gd
3       TA
4       Gd
        ..
1455    Gd
1456    Gd
1457    TA
1458    TA
1459    TA
Name: BsmtQual, Length: 1460, dtype: object

In [25]:
train['BsmtQual'].fillna('Nb') # didn't work

0       Gd
1       Gd
2       Gd
3       TA
4       Gd
        ..
1455    Gd
1456    Gd
1457    TA
1458    TA
1459    TA
Name: BsmtQual, Length: 1460, dtype: object

In [28]:
train['BsmtQual'].value_counts(), train['BsmtQual'].value_counts().sum()

(TA    649
 Gd    618
 Ex    121
 Fa     35
 Name: BsmtQual, dtype: int64,
 1423)

In [29]:
# To check if this column is affected or not by the above replacement
train['GarageCond'].value_counts(), train['GarageCond'].value_counts().sum()

(TA    1326
 Fa      35
 Gd       9
 Po       7
 Ex       2
 Name: GarageCond, dtype: int64,
 1379)

In [ ]:
train['BsmtCond'].value_counts(), train['BsmtCond'].value_counts().sum()

BsmtCond: Evaluates the general condition of the basement

       Ex	Excellent
       Gd	Good
       TA	Typical - slight dampness allowed
       Fa	Fair - dampness or some cracking or settling
       Po	Poor - Severe cracking, settling, or wetness
       NA	No Basement

With the same reason as 'BsmtQual', replace it to 'Nb (No basement)'

In [ ]:

train['BsmtCond'].value_counts(), train['BsmtCond'].value_counts().sum()

In [ ]:
train['BsmtFinType1'].value_counts(), train['BsmtFinType1'].value_counts().sum()

BsmtFinType1: Rating of basement finished area

       GLQ	Good Living Quarters
       ALQ	Average Living Quarters
       BLQ	Below Average Living Quarters	
       Rec	Average Rec Room
       LwQ	Low Quality
       Unf	Unfinshed
       NA	No Basement

With the same reason as 'BsmtQual', replace it to 'Nb (No basement)'

In [ ]:

train['BsmtFinType1'].value_counts(), train['BsmtFinType1'].value_counts().sum()

In [ ]:
train['BsmtExposure'].value_counts(), train['BsmtExposure'].value_counts().sum()

BsmtExposure: Refers to walkout or garden level walls

       Gd	Good Exposure
       Av	Average Exposure (split levels or foyers typically score average or above)	
       Mn	Mimimum Exposure
       No	No Exposure
       NA	No Basement

With the same reason as 'BsmtQual', replace it to 'Nb (No basement)'

In [ ]:

train['BsmtExposure'].value_counts(), train['BsmtExposure'].value_counts().sum()

In [ ]:
train['BsmtFinType2'].value_counts(), train['BsmtFinType2'].value_counts().sum()

BsmtFinType2: Rating of basement finished area (if multiple types)

       GLQ	Good Living Quarters
       ALQ	Average Living Quarters
       BLQ	Below Average Living Quarters	
       Rec	Average Rec Room
       LwQ	Low Quality
       Unf	Unfinshed
       NA	No Basement

With the same reason as 'BsmtQual', replace it to 'Nb (No basement)'

In [ ]:

train['BsmtFinType2'].value_counts(), train['BsmtFinType2'].value_counts().sum()

In [11]:
train['GarageCond'].value_counts(), train['GarageCond'].value_counts().sum()

(TA    1326
 Fa      35
 Gd       9
 Po       7
 Ex       2
 Name: GarageCond, dtype: int64,
 1379)

GarageCond: Garage condition

       Ex	Excellent
       Gd	Good
       TA	Typical/Average
       Fa	Fair
       Po	Poor
       NA	No Garage

In [ ]:
train['GarageQual'].value_counts(), train['GarageQual'].value_counts().sum()

In [ ]:
train['GarageFinish'].value_counts(), train['GarageFinish'].value_counts().sum()

In [ ]:
train['GarageType'].value_counts(), train['GarageType'].value_counts().sum()

In [ ]:
train['GarageYrBlt'].value_counts(), train['GarageYrBlt'].value_counts().sum()

In [ ]:
train['LotFrontage'].value_counts(), train['LotFrontage'].value_counts().sum()

In [ ]:
train['FireplaceQu'].value_counts(), train['FireplaceQu'].value_counts().sum()

In [ ]:
train['Fence'].value_counts(), train['Fence'].value_counts().sum()

In [ ]:
train['Alley'].value_counts(), train['Alley'].value_counts().sum()

In [ ]:
train['MiscFeature'].value_counts(), train['MiscFeature'].value_counts().sum()

In [ ]:
train['PoolQC'].value_counts(), train['PoolQC'].value_counts().sum()

In [ ]:
missing.loc[missing['count']>0].sort_values(by='count')